src/test/java/org/broadinstitute/hellbender/utils/hmm/HMMUnitTest.java

```diff
@@ -0,0 +1,542 @@
+package org.broadinstitute.hellbender.utils.hmm;
+
+import org.apache.commons.math3.util.Pair;
+import org.broadinstitute.hellbender.utils.R.RScriptExecutor;
+import org.broadinstitute.hellbender.utils.test.BaseTest;
+import org.broadinstitute.hellbender.utils.tsv.DataLine;
+import org.broadinstitute.hellbender.utils.tsv.TableReader;
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Unit test for HMM api classes including {@link ViterbiAlgorithm}, {@link ForwardBackwardAlgorithm} and
+ * {@link ForwardBackwardAlgorithm.Result}.
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public final class HMMUnitTest extends BaseTest {
```

Comment:
A simpler and perfectly adequate version of that last one is that if the initial probabilities vector is v and the (position-dependent) transition matrices are M_1, M_2. . . then the probabilities vector at position k is M_k M_{k-1} . . . M_2 M_1 v.